apps/astarte_pairing/lib/astarte_pairing/to0_util.ex

```diff
@@ -41,4 +43,159 @@ defmodule Astarte.Pairing.TO0Util do
         {:error, {:cbor_decode_error, reason}}
     end
   end
+
+  def get_ownership_voucher() do
+    MockFDOApi.get_ownership_voucher()
+  end
+
+  def decode_ownership_voucher(voucher_pem) do
+    ov_data =
+      voucher_pem
+      |> String.replace("-----BEGIN OWNERSHIP VOUCHER-----", "")
+      |> String.replace("-----END OWNERSHIP VOUCHER-----", "")
+      |> String.replace(~r/\s/, "")
+
+    with {:ok, cbor_data} <- Base.decode64(ov_data),
+         {:ok, result, _rest} <- CBOR.decode(cbor_data) do
+      {:ok, result}
+    else
+      {:error, reason} -> {:error, reason}
+      other -> {:error, "PEM parsing failed: #{inspect(other)}"}
+    end
+  end
+
+  def get_owner_private_key() do
+    MockFDOApi.get_owner_private_key()
+  end
+
+  def sign_with_owner_key(data, owner_key) do
+    case :public_key.pem_decode(owner_key) do
+      [{:ECPrivateKey, der_data, :not_encrypted}] ->
+        with {:ok, ec_private_key} <- safe_der_decode(der_data),
+             {:ok, raw_priv} <- extract_raw_private_key(ec_private_key),
+             {:ok, signature} <- safe_sign(data, raw_priv) do
+          {:ok, signature}
+        else
+          {:error, reason} ->
+            Logger.error("Signing failed", error: inspect(reason))
+            {:error, reason}
+        end
+
+      _ ->
+        Logger.error("PEM decode failed for owner key")
+        {:error, :pem_decode_failed}
+    end
+  end
+
+  def safe_der_decode(der_data) do
+    try do
+      {:ok, :public_key.der_decode(:ECPrivateKey, der_data)}
+    rescue
+      e -> {:error, {:der_decode_failed, e}}
+    end
+  end
+
+  defp extract_raw_private_key({:ECPrivateKey, _version, bin, _params, _pub, _extra}) do
+    case byte_size(bin) do
+      32 -> {:ok, bin}
+      n when n < 32 -> {:ok, :binary.copy(<<0>>, 32 - n) <> bin}
+      n -> {:error, {:key_too_long, n}}
+    end
+  end
+
+  def safe_sign(data, priv_key) do
+    # ECDSA with SHA-256 requires a 32-byte private key secp256r1
+    if is_binary(priv_key) and byte_size(priv_key) == 32 do
+      try do
+        {:ok, :crypto.sign(:ecdsa, :sha256, data, [priv_key, :secp256r1])}
+      rescue
+        e -> {:error, {:signing_failed, e}}
+      end
+    else
+      {:error, :invalid_private_key_format}
+    end
+  end
+
+  def build_rv_to2_addr_entry(ip, dns, port, protocol) do
+    rv_entry = [ip, dns, port, protocol]
+    encoded_entry = CBOR.encode([rv_entry])
+    {:ok, encoded_entry}
+  end
+
+  defp build_to1d_rv(entries) do
+    to1d_rv = CBOR.encode([entries])
+    {:ok, to1d_rv}
+  end
+
+  defp build_to0d(ov, wait_seconds, nonce) do
+    to0d = CBOR.encode([ov, wait_seconds, nonce])
+    {:ok, to0d}
+  end
+
+  defp add_cbor_tag(payload) do
+    %CBOR.Tag{tag: :bytes, value: payload}
+  end
+
+  defp build_to1d_to0d_hash(to0d) do
+    to1d_to0d_hash_value = :crypto.hash(:sha256, to0d)
+    # 47 is the key for SHA-256
+    to1d_to0d_hash = CBOR.encode([47, to1d_to0d_hash_value])
```

Comment:
usually it's best to have your code speak for itself rather than adding comments

```suggestion
    to1d_to0d_hash = CBOR.encode([@sha256, to1d_to0d_hash_value])
```